File: backend/app/routers/delivery.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.orm import Session, joinedload

from app.database import get_db
from app.dependencies import get_current_user, require_roles
from app.models import DeliveryTracking, Order, OrderItem, OrderStatus, User, UserRole
from app.schemas.order import DeliveryTrackingResponse, OrderResponse

router = APIRouter(prefix="/api/delivery", tags=["Delivery Tracking"])


class UpdateDeliveryStatus(BaseModel):
    status: str
    latitude: str | None = None
    longitude: str | None = None


@router.get("/track/{order_id}", response_model=DeliveryTrackingResponse)
def track_order(order_id: int, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    order = db.query(Order).filter(Order.id == order_id).first()
    if not order:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Order not found")
    if order.user_id != user.id and user.role not in (UserRole.admin, UserRole.delivery):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Access denied")
    if not order.delivery:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tracking not found")
    return order.delivery


@router.put("/track/{order_id}", response_model=DeliveryTrackingResponse)
def update_tracking(
    order_id: int,
    body: UpdateDeliveryStatus,
    user: User = Depends(require_roles(UserRole.delivery, UserRole.admin)),
    db: Session = Depends(get_db),
):
    order = db.query(Order).filter(Order.id == order_id).first()
    if not order or not order.delivery:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Order not found")
    tracking = order.delivery
    tracking.status = body.status
    tracking.latitude = body.latitude
    tracking.longitude = body.longitude
    if user.role == UserRole.delivery:
        tracking.delivery_partner_id = user.id
    status_map = {
        "out_for_delivery": OrderStatus.out_for_delivery,
        "delivered": OrderStatus.delivered,
    }
    if body.status in status_map:
        order.status = status_map[body.status]
    db.commit()
    db.refresh(tracking)
    return tracking


@router.get("/today", response_model=list[OrderResponse])
def today_deliveries(
    delivery_date: date | None = None,
    _: User = Depends(require_roles(UserRole.delivery, UserRole.admin)),
    db: Session = Depends(get_db),
):
    target = delivery_date or date.today()
    return (
        db.query(Order)
        .options(joinedload(Order.items).joinedload(OrderItem.product))
        .filter(Order.delivery_date == target, Order.status.in_([OrderStatus.confirmed, OrderStatus.out_for_delivery]))
        .order_by(Order.id)
        .all()
    )

File: src/handlers/delivery.ts
import { isResponse, requireAuth } from "../lib/auth";
import { error, json, parseJson } from "../lib/http";
import { loadOrderWithItems } from "../lib/orders";

export async function trackOrder(request: Request, env: Env, orderId: number): Promise<Response> {
	const user = await requireAuth(request, env);
	if (isResponse(user)) return user;

	const order = await env.DB.prepare<{ user_id: number }>("SELECT user_id FROM orders WHERE id = ?")
		.bind(orderId)
		.first();
	if (!order) return error("Order not found", 404);
	if (
		order.user_id !== user.id &&
		user.role !== "admin" &&
		user.role !== "delivery"
	) {
		return error("Access denied", 403);
	}

	const tracking = await env.DB.prepare("SELECT * FROM delivery_tracking WHERE order_id = ?")
		.bind(orderId)
		.first();
	if (!tracking) return error("Tracking not found", 404);
	return json(tracking);
}

export async function updateTracking(
	request: Request,
	env: Env,
	orderId: number,
): Promise<Response> {
	const user = await requireAuth(request, env, ["delivery", "admin"]);
	if (isResponse(user)) return user;

	const body = await parseJson<{
		status: string;
		latitude?: string;
		longitude?: string;
	}>(request);
	if (!body?.status) return error("status is required", 422);

	await env.DB.prepare(
		`UPDATE delivery_tracking SET status = ?, latitude = ?, longitude = ?,
     delivery_partner_id = ?, updated_at = CURRENT_TIMESTAMP WHERE order_id = ?`,
	)
		.bind(
			body.status,
			body.latitude ?? null,
			body.longitude ?? null,
			user.role === "delivery" ? user.id : null,
			orderId,
		)
		.run();

	if (body.status === "out_for_delivery") {
		await env.DB.prepare("UPDATE orders SET status = 'out_for_delivery' WHERE id = ?")
			.bind(orderId)
			.run();
	}
	if (body.status === "delivered") {
		await env.DB.prepare("UPDATE orders SET status = 'delivered' WHERE id = ?").bind(orderId).run();
	}

	const tracking = await env.DB.prepare("SELECT * FROM delivery_tracking WHERE order_id = ?")
		.bind(orderId)
		.first();
	return json(tracking);
}

export async function todayDeliveries(request: Request, env: Env): Promise<Response> {
	const user = await requireAuth(request, env, ["delivery", "admin"]);
	if (isResponse(user)) return user;

	const url = new URL(request.url);
	const deliveryDate = url.searchParams.get("delivery_date") ?? new Date().toISOString().slice(0, 10);

	const { results } = await env.DB.prepare(
		`SELECT id FROM orders WHERE delivery_date = ? AND status IN ('confirmed', 'out_for_delivery')`,
	)
		.bind(deliveryDate)
		.all();

	const orders = [];
	for (const row of results) {
		const full = await loadOrderWithItems(env.DB, row.id as number);
		if (full) orders.push(full);
	}
	return json(orders);
}
